.github/scripts/update-tags.sh

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
+
+SCRIPT="$(readlink -f "$0")"
+SCRIPTPATH="$(dirname "${SCRIPT}")"
+
+IMAGEJSON="${SCRIPTPATH}/../tests/images.json"
+
+if ! command -v crane &> /dev/null; then
+  echo Please install crane
+  exit 1
+fi
+
+if ! command -v jq &> /dev/null; then
+  echo Please install jq
+  exit 1
+fi
+
+if ! command -v yq &> /dev/null; then
+  echo Please install yq
+  exit 1
+fi
+
+if ! command -v trivy &> /dev/null; then
+  echo Please install trivy
+  exit 1
+fi
+
+if ! command -v python3 -c 'import ruamel.yaml' &> /dev/null; then
+  echo Please install python3 with the ruamel.yaml module
+  exit 1
+fi
+
+if ! command -v python3 -c 'import dict_deep' &> /dev/null; then
+  echo Please install python3 with the dict_deep module
+  exit 1
+fi
+
+jq -r '. | keys[]' "$IMAGEJSON" | while read -r CHART; do
+  jq -r ".\"${CHART}\" | keys[]" "$IMAGEJSON" | while read -r IDX; do
+    QUERY=$(jq -r ".\"${CHART}\"[${IDX}].query" "$IMAGEJSON")
+    FILTER=$(jq -r ".\"${CHART}\"[${IDX}].filter" "$IMAGEJSON")
+    SORTFLAGS=$(jq -r ".\"${CHART}\"[${IDX}].\"sort-flags\"" "$IMAGEJSON")
+
+    VALUES="${SCRIPTPATH}/../../charts/spire/charts/${CHART}"
+    REGISTRY=$(yq e ".${QUERY}.registry" "$VALUES")
+    REPOSITORY=$(yq e ".${QUERY}.repository" "$VALUES")
+    VERSION=$(yq e ".${QUERY}.tag" "$VALUES")
+    # shellcheck disable=SC2086
+    LATEST_VERSION=$(crane ls "${REGISTRY}/${REPOSITORY}" | grep "${FILTER}" | sort ${SORTFLAGS}| tail -n 1)
```

Comment:
```suggestion
    LATEST_VERSION=$(crane ls "${REGISTRY}/${REPOSITORY}" | grep "${FILTER}" | sort "${SORTFLAGS[@]}"| tail -n 1)
```

This might do the trick without disabling the linting. Not sure though in this usecase.

Reply:
No. the nonquoting is critical to its functioning. multiple args are passed through the variable and have to be handled separately by bash. The linter is trying to catch a common case of accidental misconfiguration but in this case, it really is intentional.

Reply:
Can we put in a comment on why we have to disable the linter here?

Reply:
Done

Reply:
> This might do the trick without disabling the linting. Not sure though in this usecase.

I tried it again just to be sure... the problem is the variable is a string with spaces, not an array, so the [@] form doesn't work.

Reply:
Will try.

Reply:
same issue. its trying to build one array argument out of a string intended to hold multiple:
sort: multi-character tab ‘ . -k1,1n -k2,2n -k3,3n’

I believe unquoted really is the correct usage.

You can test for yourself by setting spire-agent.waitForIt.image.tag  to latest-20220517
and running ./.github/scripts/update-tags.sh

Reply:
Both of these approaches work, and both work without shellcheck reporting an error, or needing quoting disabled.

```
SORTFLAGS=( "-k1,1n" "-k2,2n" "-k3,3n" )

sort "${SORTFLAGS[@]}" << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF

printf 'a, b, c, 1\na, c, c, 2\na, b, d, 3\n' | sort "${SORTFLAGS[@]}"
```

The idea that "${VARIABLE[@]}" expands to "-k1,1n -k2,2n -k3,3n" isn't true.  For arrays, expansion generally puts the double quotes around each element of an array expansion "-k1,1" "-k2,2" "-k3,3".

to illustrate, I use this bit of code to dump an array.  (It's from a shell library project I wrote, to manage function extensions in an "object_oriented" like manner)

```
function dump_object() {
        declare OBJECT_VAR=$1
        declare -n OBJECT_REF=$1

        SORTED_KEYS=()
        PREVIOUS_IFS=${IFS}
        while IFS= read -rd '' KEY; do
                SORTED_KEYS+=( "${KEY}" )
        done < <(printf '%s\0' "${!OBJECT_REF[@]}" | sort -z)
        for KEY in ${SORTED_KEYS[*]}; do
                echo ${OBJECT_VAR}[$KEY]=${OBJECT_REF[${KEY}]}
        done
        IFS=${PREVIOUS_IFS}
}
```

to use it one does
```
SORTFLAGS=( "-k1,1n" "-k2,2n" "-k3,3n" )
dump_object SORTFLAGS
```

to get 
```
SORTFLAGS[0]=-k1,1n
SORTFLAGS[1]=-k2,2n
SORTFLAGS[2]=-k3,3n
```

in the definition of SORTFLAGS above, the double quoting in the sort works perfectly.

```
#!/bin/bash

function dump_object() {
        declare OBJECT_VAR=$1
        declare -n OBJECT_REF=$1

        SORTED_KEYS=()
        PREVIOUS_IFS=${IFS}
        while IFS= read -rd '' KEY; do
                SORTED_KEYS+=( "${KEY}" )
        done < <(printf '%s\0' "${!OBJECT_REF[@]}" | sort -z)
        for KEY in ${SORTED_KEYS[*]}; do
                echo ${OBJECT_VAR}[$KEY]=${OBJECT_REF[${KEY}]}
        done
        IFS=${PREVIOUS_IFS}
}

SORTFLAGS=( "-k1,1n" "-k2,2n" "-k3,3n" )

dump_object SORTFLAGS

sort "${SORTFLAGS[@]}" << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF
```

but if I do
```
BAD_SORTFLAGS=( "-k1,1n -k2,2n -k3,3n" )

dump_object BAD_SORTFLAGS
```
which shows as output
```
BAD_SORTFLAGS[0]=-k1,1n -k2,2n -k3,3n
```

Then running
```
sort "${BAD_SORTFLAGS[@]}" << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF
```
or
```
printf 'a, b, c, 1\na, c, c, 2\na, b, d, 3\n' | sort "${SORTFLAGS[@]}"
```
doesn't work but the unquoted versions
```
sort ${BAD_SORTFLAGS[@]} << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF
```
and
```
printf 'a, b, c, 1\na, c, c, 2\na, b, d, 3\n' | sort ${SORTFLAGS[@]}
```
do work.

Odds are you are misidentifying the root cause.  It's not because the quotes around ${SORTFLAGS[@]} are combining the parameters, it's because you only set one value in ${SORTFLAGS[@]} that has spaces separating the text within that single value.

Please never disable the quoting rules in shellcheck, of all the rules, it's the most important one.

Reply:
The source of the string is not a bash array though, but a json document. How do you suggest we pull it into a bash array?

Reply:
Its defined in the json config like:
https://github.com/spiffe/helm-charts/pull/249/files#diff-4c6c57c7cd483f8116b836fc3762a95f0f347b5f1f5e4cb67002f247d50ff14bR24

Then pulled into bash here:
https://github.com/spiffe/helm-charts/pull/249/files#diff-e800c023ce454983c68686a85f2ea49f6114c2cd433a239eb9637c84ae3ddeefR42

There could be a better way to structure that, but no idea what that should look like